Introduction to Android Arch. Components. (#100)
rootstrap/blog: content/Android Arc Components - Amaury Ricard.md

## Introduction to Android Architecture Components

According to https://developer.android.com:
"Android architecture components are a collection of libraries that help you design robust, testable, and maintainable apps. Start with classes for managing your UI component lifecycle and handling data persistence."

Android Architecture Components are focused to solve many problems that Android developers face on a project. It focuses on a clean and easy to maintain architecture such as MVVM, and trying to take advantage of all the capabilities of the Android framework and how it is structured.

![ArchComponets](images/arch_componets.png)

This post is to give a brief introduction to Android Architecture Components, in later posts we will delve deeper into each of these libraries aiming to show the capabilities of each one.

Let's start:

- ### Life cycle:

> Android introduce lifecycle as a way to avoid memory leaks and keep a cleaner code, and each activity|fragment|lifecycle-owner can be abstracted from the behavior of an object at each stage of its life cycle, that object can subscribe to the activity|fragment lifecycle implementing the interface androidx.lifecycle in order to execute certain actions in each state:

Reviewer: it should be "Android introduces"

![LifeCycle](images/lifecycle.jpg)

To subscribe to a lifecycle state you can use annotations like:

``` kotlin
  @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
  fun yourOnResumeActionListener() {
	...	
  }
```

Eache states:
ON_RESUME,ON_START,ON_STOP,ON_DESTROY,ON_PAUSE,ON_CREATE
Note: Each activity|fragment|lifecycle-owner needs to add the object to his own lifecycle observers:

``` kotlin 
    this.getLifecycle().addObserver(objectWithLifeCycleListener)
```

- ### Live Data:

> This concept is not new for android, this is a new way to see and observable object that was previously introduced with "androidx". With live data they introduce it a owner in order to know when to notify data depending of the owner lifecycle in order to manage consistent data and avid memory leaks. 

Reviewer: "introduce it TO AN owner", and also "data depending ON the owner". In the last line, it should be "avoid", it's missing an O.

A classic example: for an app it is important to keep the user data updated, if the user changes the name, avatar, permissions, etc... you can observe it and notify the app for those changes.

- ### View Model:

> Is the one with the responsibility of provide and keep data save for the UI in each state of the lifecycle of the view or the app for example screen rotations or when the app gone to background or is sleeping, is the one who "talk" with the repository subscribe to  the repository response with LiveData, each view(Activity|Fragment not custom views or Adapters) should have a view model example:

Reviewer: It should be "IT'S the one with the responsibility of providING and keepING data saveD", and then on the second line, "or when the app HAS gone to background". Also, "IT'S the one who "talkS" TO the repository subscribeD", and a space needs to be added after "each view".

MainActivity.kt
MainActivityViewModel.kt

> Nota that you needs to setup the view model with the View in order to register the view as the LifeCycle owner:

Reviewer: "NotE that you NEED" (minus the S)

> In the view(this case MainActivity):

Reviewer: add space after view

``` kotlin
val viewModel = ViewModelProviders.of(this)
    .get(MainActivityViewModel::class.java)
```

- ### Room database:

> Android always use Sqlite to storage|manage data but the use of Sqlite required write a lot of code or the use of unsafety third party libraries, also Sqlite didn't check the sql queries at compile time, they created Room to solve that problem and keep data safe and integrated it with LiveData to provide a clean wait to provide data to the ViewModel, to create a db entity all you have to do is add @Entity to a Model class and a create a @Dao interface to manage that entity values:

Reviewer: "always useS", also "required TO write a lot of code", and it should be "unsafe third party libraries".

``` kotlin
@Entity
class Person {
    @PrimaryKey
    var id: Long? = null
    var name: String? = null
    var lastName: Long? = null
}

@Dao
interface PersonDAO {
    @Insert( onConflict = OnConflictStrategy.REPLACE )
    fun insertPerson(person: Person): Long
 
    @Query("SELECT * FROM person WHERE id = :arg0")
    fun findById(id: Long): LiveData<Person>
}

// Setup the DB:
@Database( entities = arrayOf(Person::class), version = 1)
abstract class Database : RoomDatabase() {
    abstract fun personDAO(): PersonDAO
}
```

- ### Managing your UI components:

Android introduced a Data Binding library to provide us a new way to set up the view, as passing all data or livedata to the xml file:

``` xml
<layout xmlns:android="http://schemas.android.com/apk/res/android"
        xmlns:app="http://schemas.android.com/apk/res-auto">
    <data>
        <variable
            name="viewmodel"
            type="com.rootstrap.viewmodel.ProfileActivityViewModel" />
    </data>
    <ConstraintLayout ...>
        <TextView
            ...
            android:text="@{viewmodel.profile.name}" />
  </ConstraintLayout>
</layout>
```

> Also the added binding adapters in order to create custom properties or listeners for each layout for example:

Reviewer: "Also theY added"

![ArchComponets](images/arch_componets.png)

This post is to give a brief introduction to Android Architecture Components, in later posts we will delve deeper into each of these libraries aiming to show the capabilities of each one.

Let's start:

- ### Life cycle:

Android introduces lifecycle as a way to avoid memory leaks and keep a cleaner code, and each activity|fragment|lifecycle-owner can be abstracted from the behavior of an object at each stage of its life cycle, that object can subscribe to the activity|fragment lifecycle implementing the interface androidx.lifecycle in order to execute certain actions in each state:

![LifeCycle](images/lifecycle.jpg)

To subscribe to a lifecycle state you can use annotations like:

``` kotlin
  @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
  fun yourOnResumeActionListener() {
	...	
  }
```

Eache states:
ON_RESUME,ON_START,ON_STOP,ON_DESTROY,ON_PAUSE,ON_CREATE
Note: Each activity|fragment|lifecycle-owner needs to add the object to his own lifecycle observers:

``` kotlin 
    this.getLifecycle().addObserver(objectWithLifeCycleListener)
```

- ### Live Data:

> This concept is not new for android, this is a new way to see and observable object that was previously introduced with "androidx". With live data they introduce it to an owner in order to know when to notify data depending on the owner lifecycle in order to manage consistent data and avoid memory leaks. 

Author: 😂💪

> A classic example: for an app it is important to keep the user data updated, if the user changes the name, avatar, permissions, etc... you can observe it and notify the app for those changes.

Author: Agree but this could a subject for another post.

``` kotlin
@Entity
class Person {
    @PrimaryKey
    var id: Long? = null
    var name: String? = null
    var lastName: Long? = null
}

@Dao
interface PersonDAO {
    @Insert( onConflict = OnConflictStrategy.REPLACE )
    fun insertPerson(person: Person): Long
 
    @Query("SELECT * FROM person WHERE id = :arg0")
    fun findById(id: Long): LiveData<Person>
}

// Setup the DB:
@Database( entities = arrayOf(Person::class), version = 1)
abstract class Database : RoomDatabase() {
    abstract fun personDAO(): PersonDAO
}
```

- ### Managing your UI components:

Android introduced a Data Binding library to provide us a new way to set up the view, as passing all data or livedata to the xml file:

``` xml
<layout xmlns:android="http://schemas.android.com/apk/res/android"
        xmlns:app="http://schemas.android.com/apk/res-auto">
    <data>
        <variable
            name="viewmodel"
            type="com.rootstrap.viewmodel.ProfileActivityViewModel" />
    </data>
    <ConstraintLayout ...>
        <TextView
            ...
            android:text="@{viewmodel.profile.name}" />
  </ConstraintLayout>

> </layout>

Reviewer 2: This is awesome!

```

> Also they added binding adapters in order to create custom properties or listeners for each layout for example:

Reviewer 2: ```suggestion
Also, they added binding adapters to create custom properties or listeners for each layout. For example:
```

``` kotlin
@BindingAdapter("app:addPersonAdapter")
fun personAdapter(view: RecyclerView, adapter: PersonAdapter) {
	view.apply {
	    it.layoutManager = LinearLayoutManager()
	    It.adapter = adapter

>     }

Reviewer 2: Indentation. And the `it` on the second line should not be capitalized.

}
```

The view:

``` xml
    ...
    <variable
            name="viewmodel"
            type="com.rootstrap.viewmodel.ProfileActivityViewModel" />
    </variable>
    ...
    <RecyclerView
        ...
        add:addPersonAdapter="@{viewmodel.personAdapter}" />
```

> Also we can integrate DataBinding with Live data but that's for another post.

Reviewer 2: ```suggestion
Also, we can integrate DataBinding with LiveData, but that's for another post.
```

- ### Conclusions: 

> As we have observed, these libraries come to help us obtain a cleaner, faster, more scalable and easier to maintain code, as I mentioned earlier, in later posts we will delve deeper into each of these components in order to provide a better understanding of each one.

Reviewer 2: ```suggestion
As we have observed, these libraries come to help us write code that's cleaner, faster, more scalable, and easier to maintain. As I mentioned earlier, in later posts we will delve deeper into each of these components to provide a better understanding.
```

## Introduction to Android Architecture Components

> According to https://developer.android.com:

Reviewer 3: According to [Android Developers](https://developer.android.com/)

> "Android architecture components are a collection of libraries that help you design robust, testable, and maintainable apps. Start with classes for managing your UI component lifecycle and handling data persistence."

Author: That's literally what they have in the developer's page regarding arch. components:
https://developer.android.com/topic/libraries/architecture

> "Android architecture components are a collection of libraries that help you design robust, testable, and maintainable apps. Start with classes for managing your UI component lifecycle and handling data persistence."

Reviewer 2: Yeah, but I think that that sentence only makes sense with the bullet points that follow. If I read that I would think: "oh, ok, I'll start with that, but then what?" 😂

> Android Architecture Components are focused to solve many problems that Android developers face on a project. It focuses on a clean and easy to maintain architecture such as MVVM, and trying to take advantage of all the capabilities of the Android framework and how it is structured.

Reviewer 3: ```suggestion
Android Architecture Components help solve many problems that Android developers face on a project. It helps create a clean and easy to maintain architecture such as MVVM, with the goal of taking advantage of all the capabilities of the Android framework and it's structure.
```

> ![ArchComponets](images/arch_componets.png)

Reviewer 4: @amaury901130 There's a type in `arch_componets` here, missing an `n`

> This post is to give a brief introduction to Android Architecture Components. In later posts, we will delve deeper into each of these libraries aiming to show the capabilities of each one.

Reviewer 3: ```suggestion
This post will give a brief introduction to Android Architecture Components. In later posts, we will delve deeper into each of the libraries to show the capabilities of each one.
```

Let's start:

> - ### Life cycle:

Reviewer 3: ```suggestion
- ### Android Lifecycle:
```

> This post is to give a brief introduction to Android Architecture Components, in later posts we will delve deeper into each of these libraries aiming to show the capabilities of each one.

Reviewer 2: ```suggestion
This post is to give a brief introduction to Android Architecture Components. In later posts, we will delve deeper into each of these libraries aiming to show the capabilities of each one.
```

> Android introduces lifecycle as a way to avoid memory leaks and keep a cleaner code, and each activity|fragment|lifecycle-owner can be abstracted from the behavior of an object at each stage of its life cycle, that object can subscribe to the activity|fragment lifecycle implementing the interface androidx.lifecycle in order to execute certain actions in each state:

Reviewer 2: ```suggestion
Android introduces lifecycle as a way to avoid memory leaks and keep a cleaner code, and each activity|fragment|lifecycle-owner can be abstracted from the behavior of an object at each stage of its life cycle. That object can subscribe to the activity|fragment lifecycle implementing the interface androidx.lifecycle in order to execute certain actions in each state:
```

![LifeCycle](images/lifecycle.jpg)

To subscribe to a lifecycle event you can use annotations like:

``` kotlin
  @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
  fun yourOnResumeActionListener() {
	...	
  }
```

Available states:
- ON_RESUME
- ON_START
- ON_STOP
- ON_DESTROY
- ON_PAUSE
- ON_CREATE

Note: Each activity|fragment|lifecycle-owner needs to add the object to its own lifecycle observers:

``` kotlin 
    this.getLifecycle().addObserver(objectWithLifeCycleListener)
```

- ### Live Data:

LiveData is a new type of Observable that is lifecycle-aware. This means that LiveData takes into account their observer's lifecycle state to notify the changes to them, meaning that only active components (STARTED or RESUMED) will be notified, avoiding potential issues such as Memory Leaks, crashes due to stopped activities and dealing with configuration changes properly. All of this while keeping up to date data everywhere in your app. Isn't it awesome?

A classic example: for an app it is important to keep the user data updated, if the user changes the name, avatar, permissions, etc... you can observe it and notify the app for those changes.

- ### View Model:

> It's the one with responsible for providing and keeping data saved for the UI in each state of the lifecycle of the view or the app, for example: screen rotations; when the app has gone to background; when is sleeping. 

Reviewer 2: ```suggestion
It's the one responsible for providing and keeping data saved for the UI in each state of the lifecycle of the view or the app. For example: screen rotations; when the app has gone to the background; when it's sleeping. 
```

> It is the one who "talks" to the repository and subscribed to the repository response with LiveData, 

Reviewer 2: ```suggestion
It is the one who "talks" to the repository and it's subscribed to its response using LiveData.
```

> Each view (Activity|Fragment not custom views or Adapters) should have a view model example:

Reviewer 2: ```suggestion
Each view (Activity|Fragment not custom views or Adapters) should have a view model. Example:
```

    var name: String? = null
    var lastName: Long? = null
}

@Dao
interface PersonDAO {
    @Insert( onConflict = OnConflictStrategy.REPLACE )
    fun insertPerson(person: Person): Long
 
    @Query("SELECT * FROM person WHERE id = :arg0")
    fun findById(id: Long): LiveData<Person>
}

// Setup the DB:
@Database( entities = arrayOf(Person::class), version = 1)
abstract class Database : RoomDatabase() {
    abstract fun personDAO(): PersonDAO
}
```

- ### Managing your UI components:

Android introduced a Data Binding library to provide us a new way to set up the view, as passing all data or livedata to the xml file:

``` xml
<layout xmlns:android="http://schemas.android.com/apk/res/android"
        xmlns:app="http://schemas.android.com/apk/res-auto">
    <data>
        <variable
            name="viewmodel"
            type="com.rootstrap.viewmodel.ProfileActivityViewModel" />
    </data>
    <ConstraintLayout ...>
        <TextView
            ...
            android:text="@{viewmodel.profile.name}" />
  </ConstraintLayout>
</layout>
```

Also, they added binding adapters to create custom properties or listeners for each layout. For example:

``` kotlin

> @BindingAdapter("app:addPersonAdapter")
> fun personAdapter(view: RecyclerView, adapter: PersonAdapter) {
>   view.apply {
>     it.layoutManager = LinearLayoutManager()
> 	it.adapter = adapter
>   }
> }

Reviewer 2: ```suggestion
@BindingAdapter("app:addPersonAdapter")
fun personAdapter(view: RecyclerView, adapter: PersonAdapter) {
  view.apply {
    it.layoutManager = LinearLayoutManager()
    it.adapter = adapter
  }
}
```

```

The view:

``` xml
    ...
    <variable
            name="viewmodel"
            type="com.rootstrap.viewmodel.ProfileActivityViewModel" />
    </variable>
    ...
    <RecyclerView
        ...
        add:addPersonAdapter="@{viewmodel.personAdapter}" />
```

Also, we can integrate DataBinding with LiveData, but that's for another post.

- ### Conclusions: 

> As we have observed, these libraries come to help us write code that's cleaner, faster, more scalable, and easier to maintain. As I mentioned earlier, in later posts we will delve deeper into each of these components to provide a better understanding.

Reviewer 3: ```suggestion
As we have observed, these libraries help us write code that's cleaner, faster, more scalable, and easier to maintain.  In future posts, we will delve deeper into each of these components to provide a better understanding.
```
